### PR TITLE
Enhanced persistence priority rate limiting

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Take caller and call initiation into consideration when performing persistence rate limiting.
- Caller is either `system` or user namespace name
- Call initiation is only defined when caller type is `api` and its value will be the frontend method name
- If caller is user namespace name (not `system`), the persistence call will first consume a per ns token and then a host level token. 
- Specified caller and call origin information

**Why?**
- System protection

**How did you test it?**
- Will add more unit test.
- Will test locally.

**Potential risks**

**Is hotfix candidate?**
- No

### CODE REVIEW

File: common/headers/caller_info.go [hunk 1]
```diff
@@ -33,47 +33,117 @@ import (
...
+		CallerName: CallerNameSystem,
+		CallerType: CallerTypeBackground,
+	}
+)
 
-	// TODO: add fields for CallerName and CallerInitiation
+type CallerInfo struct {
+	CallerName     string
+	CallerType     string
+	CallInitiation string
```
Comment (yiminc) [inline common/headers/caller_info.go:50]: nit: name suggestion CallOrigin

File: common/headers/caller_info.go [hunk 2]
```diff
@@ -33,47 +33,117 @@ import (
...
+	SystemBackgroundCallerInfo CallerInfo = CallerInfo{
+		CallerName: CallerNameSystem,
+		CallerType: CallerTypeBackground,
+	}
+)
 
-	// TODO: add fields for CallerName and CallerInitiation
+type CallerInfo struct {
+	CallerName     string
+	CallerType     string
```
Comment (yiminc) [inline common/headers/caller_info.go:49]: add comments to indicate CallerName could be system or user namespace, caller type could be API or background.

File: common/persistence/client/quotas.go [hunk 3]
```diff
@@ -30,23 +30,57 @@ import (
...
 
-	APIPriorityOverride = map[string]int{
+	APITypeCallInitiationPriorityOverride = map[string]int{
+		"StartWorkflowExecution":           0,
+		"SignalWithStartWorkflowExecution": 0,
+		"SignalWorkflowExecution":          0,
+		"RequestCancelWorkflowExecution":   0,
+		"TerminateWorkflowExecution":       0,
+		"GetWorkflowExecutionHistory":      0,
+		"UpdateWorkflow":                   0,
```
Comment (yiminc) [inline common/persistence/client/quotas.go:45]: I understand the top 3 ones, but why special treat these 4?

File: common/persistence/client/quotas.go [hunk 4]
```diff
@@ -30,23 +30,57 @@ import (
...
 func NewPriorityRateLimiter(
+	namespaceMaxQPS PersistenceNamespaceMaxQps,
+	hostMaxQPS PersistenceMaxQps,
+) quotas.RequestRateLimiter {
+	hostRequestRateLimiter := newPriorityRateLimiter(
+		func() float64 { return float64(hostMaxQPS()) },
+	)
+
+	return quotas.NewNamespaceRateLimiter(func(req quotas.Request) quotas.RequestRateLimiter {
+		if req.Caller != "" && req.Caller != headers.CallerNameSystem {
```
Comment (yiminc) [inline common/persistence/client/quotas.go:68]: this means system will have higher priority compare to user namespace

File: service/matching/taskReader.go [hunk 5]
```diff
@@ -103,7 +103,7 @@ func (tr *taskReader) Signal() {
 }
 
 func (tr *taskReader) dispatchBufferedTasks(ctx context.Context) error {
-	ctx = headers.SetCallerInfo(ctx, headers.NewCallerInfo(headers.CallerTypeBackground))
+	ctx = tr.initContext(ctx)
```
Comment (yiminc) [inline service/matching/taskReader.go:106]: same change need to be applied to db_task_reader and db_task_writer. cc @maoyi16

File: common/persistence/client/quotas.go (hunk 3 above)
Comment (yycptt) [reply common/persistence/client/quotas.go:45]: This is copied from the frontend api priority definition, which I guess there were some discussion when that code is written?
https://github.com/temporalio/temporal/blob/master/service/frontend/configs/quotas.go#L41

I agree that the top 3 are the most important ones. Let me know your thoughts if the rest 4 should be removed.

File: common/persistence/client/quotas.go (hunk 4 above)
Comment (yycptt) [reply common/persistence/client/quotas.go:68]: I am not following. System calls go through one ratelimiter. User namespace calls first go through the a per ns ratelimiter, and then the same ratelimiter as the system calls. So as long as user calls are within their limit, I don't see why they will get lower priority.

File: service/matching/taskReader.go (hunk 5 above)
Comment (yycptt) [reply service/matching/taskReader.go:116]: Done.

**Review by yiminc — APPROVED** [review]

**Pull request merged**